hydroflow/src/scheduled/handoff/handoff_list.rs

```diff
@@ -1,254 +1,159 @@
-use std::cell::Cell;
-use std::marker::PhantomData;
-use std::rc::Rc;
-
 use ref_cast::RefCast;
 use sealed::sealed;
 
-use crate::scheduled::ctx::{InputPort, OutputPort, RecvCtx, SendCtx};
 use crate::scheduled::graph::HandoffData;
+use crate::scheduled::port::{BaseCtx, BasePort};
 use crate::scheduled::type_list::TypeList;
 use crate::scheduled::{HandoffId, SubgraphId};
 
 use super::Handoff;
 
-/// A variadic list of Handoff types, represented using a lisp-style tuple structure.
-///
-/// This trait is sealed and not meant to be implemented or used directly. Instead tuple lists (which already implement this trait) should be used, for example:
-/// ```ignore
-/// type MyHandoffList = (VecHandoff<usize>, (VecHandoff<String>, (TeeingHandoff<u32>, ())));
-/// ```
-/// The [`tl!`] (tuple list) macro simplifies usage of this kind:
-/// ```ignore
-/// type MyHandoffList = tl!(VecHandoff<usize>, VecHandoff<String>, TeeingHandoff<u32>);
-/// ```
 #[sealed]
-pub trait HandoffList: TypeList {
-    type InputHid: TypeList;
-    type InputPort: TypeList;
-    type RecvCtx<'a>: TypeList;
-    fn make_input(sg_id: SubgraphId) -> (Self::InputHid, Self::InputPort);
-    fn make_recv<'a>(handoffs: &'a [HandoffData], input_hids: &Self::InputHid)
-        -> Self::RecvCtx<'a>;
-
-    type OutputHid: TypeList;
-    type OutputPort: TypeList;
-    type SendCtx<'a>: TypeList;
-    fn make_output(sg_id: SubgraphId) -> (Self::OutputHid, Self::OutputPort);
-    fn make_send<'a>(
-        handoffs: &'a [HandoffData],
-        output_hids: &Self::OutputHid,
-    ) -> Self::SendCtx<'a>;
+pub trait SendPortList: BasePortList<true> {}
+#[sealed]
+impl<T: BasePortList<true>> SendPortList for T {}
+
+#[sealed]
+pub trait RecvPortList: BasePortList<false> {}
+#[sealed]
+impl<T: BasePortList<false>> RecvPortList for T {}
+
+#[sealed]
+pub trait BasePortList<const S: bool>: TypeList {
```

Comment:
is there a way we can have something other than a `bool` here? I don't really understand what it's doing, but could it be possible to use a trait instead? Is there a reason having a trait with a bool method marked `#[inline]` is less expressive than a bool?

Reply:
Basically we have `InputPort` and `OutputPort` which in terms of implementation are pretty much identically, but I want to treat them as different types since they're not interchangeably semantically. So here we just use one struct to represent both with a `bool` to differentiate (true = send, false = recv).

But I want to change this to not be a bool, and be a different tag struct instead, because the bool isn't descriptive